### ISSUE TITLE
Potential premature dispose of HTTP/2 connection in ClientWebSocket

### ISSUE THREAD

**Issue by CarnaViire** [dotnet/runtime#74415]
I've found very suspicious code while looking at ClientWebSocket, I didn't try to create a repro yet, but it should be easy to do so.

**TL;DR:** I suspect if you try to use WebSocket with custom security options like Credentials or RemoteCertificateValidationCallback, it will result in premature HTTP/2 connection disposal, so the WebSocket will not be usable.

The code that decides whether WS should dispose the message handler after use does that regardless of HTTP version

https://github.com/dotnet/runtime/blob/3824cb2931453d7818c6fff0762cc6f1c536a147/src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs#L240-L247
https://github.com/dotnet/runtime/blob/3824cb2931453d7818c6fff0762cc6f1c536a147/src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs#L264-L266

Disposing the handler after the Upgrage request is totally ok in HTTP/1.1 case, because the connection is removed from the HttpConnectionPool. But for HTTP/2 case, the connection stays in the pool, as only one stream is used for a WebSocket and other streams can be used for HTTP requests.

So, as the code follows, after an Extended Connect call and creation of a WebSocket, the handler will get disposed, which will result in HTTP/2 connections in the pool being disposed, which will result in a closed HTTP/2 stream and an aborted WebSocket -- unless I miss something.

Could you please check out this scenario @greenEkatherine?

**Comment by MihaZupan** [comment]
Even with a shared handler instance, there is an issue that we are setting the pooled connection lifetime to zero:
https://github.com/dotnet/runtime/blob/3824cb2931453d7818c6fff0762cc6f1c536a147/src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs#L253

As far as I'm reading the code, this means that if the user just sets the version to 2.0 without passing in their own handler, they will never reuse an H2 connection for multiple WS streams.

**Comment by greenEkatherine** [comment]
I can confirm that if we don't provide external invoker but provide custom options the handler will be disposed. Should we keep connection if it wasn't provided by user (which means to disable disposing for H2 at all)?

**Comment by CarnaViire** [comment]
Yes, I think if we end up using H2 with custom security options for the WebSocket, we should keep the handler in the instance and dispose upon WS disposal.

**Comment by karelz** [comment]
Triage: We don't believe it has functional impact (@greenEkatherine please confirm).
However, we will not reuse the HTTP/2 connection for multiple WebSocket streams, which defeats the purpose of the feature.
Does not seem to be critical for 7.0 as it is new protocol.

**Comment by greenEkatherine** [comment]
Confirming that HTTP/2 stream is not closed after handler disposal -- no functional impact.

Currently there are no such scenarios in our tests, even for HTTP/1.1. We test some options, but we don't check web socket state or successful send/receive after connection.

**Comment by BrennanConroy** [comment]
SignalR is one customer of this feature, but it isn't affected because we will pass our own invoker in, and this issue only occurs when you don't pass one in.

**Comment by stephentoub** [comment]
ok.  If this is only about when no handler is passed,  I'm not even convinced this is a bug. We're not pooling connections for 1.1; we needn't for 2.0 either.  And if we do pool with an implicit ambient handler, we're taking the resource management out of the user's hands.  I'd argue pooling here should be opt-in by the user passing in the handler.

**Comment by MihaZupan** [comment]
#74473 is essentially making pooling opt-out. In my view the main (if not only) point of using WebSockets over H2 is the option to reuse a connection.
We can't reuse/pool connections for 1.1, but for H2 we can at least make it work for options compatible with a shared handler. If we wanted to we technically could expand that to other options as well in the future.

**Comment by MihaZupan** [comment]
Reopening for 7.0 backport

**Comment by karelz** [comment]
Triage: In team discussion on Thu 8/31 we decided to throw for HTTP/2 WebSockets when invoker is not passed.
We can decide in future .NET versions to change it (e.g. to pool connections - though that is going in the undesirable direction of `HttpWebRequest`/`ServicePointManager` design).

Doing nothing was also an option, though it brings only minimal value to the customers as there is no sharing of HTTP/2 connection, therefore no perf benefit over HTTP/1.1 WebSockets.
We also considered Roslyn Analyzer instead, but given that this is new feature in 7.0, we decided to throw instead while we can. We would not be able to make such change in future.

Note: This is technically a breaking change against 7.0 RC1 which is go-live, so we will have to document it as breaking change.
